instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module ActiveJob
+      module Patches
+        # Module to prepend to ActiveJob::Base for instrumentation.
+        module ActiveJobCallbacks
+          def self.prepended(base) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength, Metrics/CyclomaticComplexity
+            base.class_eval do # rubocop:disable Metrics/BlockLength
+              around_enqueue do |job, block|
+                span_kind = job.class.queue_adapter_name == 'inline' ? :client : :producer
+                span_name = "#{otel_config[:job_class_span_names] ? job.class : job.queue_name} send"
+                span_attributes = job_attributes(job)
+                otel_tracer.in_span(span_name, attributes: span_attributes, kind: span_kind) do
+                  OpenTelemetry.propagation.inject(job.metadata)
+                  block.call
+                end
+              end
+
+              around_perform do |job, block|
+                span_kind = job.class.queue_adapter_name == 'inline' ? :server : :consumer
+                span_name = "#{otel_config[:job_class_span_names] ? job.class : job.queue_name} process"
+                span_attributes = job_attributes(job).merge('messaging.operation' => 'process')
+
+                propagated_context = OpenTelemetry.propagation.extract(job.metadata)
+
+                if propagation_style == :link
+                  span_links = [
+                    OpenTelemetry::Trace::Link.new(OpenTelemetry::Trace.current_span(propagated_context).context)
+                  ]
+                end
+
+                parent_context = (propagation_style == :child ? propagated_context : OpenTelemetry::Context.current)
+                OpenTelemetry::Context.with_current(parent_context) do
+                  otel_tracer.in_span(span_name, attributes: span_attributes, links: span_links, kind: span_kind) do |span|
+                    span.set_attribute('messaging.active_job.executions', job.executions)
+                    block.call
+                  end
+                end
+              ensure
+                # We may be in a job system (eg: resque) that forks and kills worker processes often.
+                # We don't want to lose spans by not flushing any span processors, so we force it here.
+                otel_tracer.tracer_provider.force_flush
```

Comment:
This is an interesting addition I want to think about a little.  I have it a backlog todo list to add shutdown hooks to sidekiq as part of the instrumentation so every application owner isn't responsible for adding at as part of its config.  

I wonder if we should do the same for resque instead of pushing the responsibility of it onto a place like this.

Reply:
Barring my above comment I'm wonder if we should consider this type of scenario `absolutely necessary`.  

>ForceFlush SHOULD only be called in cases where it is absolutely necessary, such as when using some FaaS providers that may suspend the process after an invocation, but before the SpanProcessor exports the completed spans.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#forceflush-1

Reply:
> Barring my above comment I'm wonder if we should consider this type of scenario absolutely necessary.

Yes, actually I would say this would qualify as `absolutely necessary`. If you don't flush, you run the very real chance of resque (or another forking job processor) killing things before the export deadline rolls around for a `BatchSpanProcessor`. Maybe even worth updating the spec to mention it.

In a really ideal world we could add an exit hook for `resque` to flush rather than doing it after a job is processed in ActiveJob, but we'd also want to do that for other queues as well (sidekiq does this kind of thing too, I think?). Maybe carrying patches for just those two would be okay in this instrumentation, but I would start to feel rather odd doing it for more than just one or two queueing systems. Feels like a lot of work... but in a bad way - more like "not sure if that's going to cause us headaches or not later one".

I don't think we could really rely on the sidekiq or resque instrumentation adding necessary shutdown hooks either; there's no real guarantee that someone loaded those instrumentations.

If we decide that this is not a good idea, we could possibly make it optional - either an instrumentation option, or maybe check for an env variable or something.